ql/src/java/org/apache/hadoop/hive/ql/exec/vector/ptf/VectorPTFGroupBatches.java

```diff
@@ -858,6 +858,15 @@ private void runEvaluatorForRow(int evaluatorIndex, VectorPTFEvaluatorBase evalu
 
     Object result = null;
     if (evaluator.canRunOptimizedCalculation(rowNum, range)) {
+      /*
+       * A classic evaluator (which doesn't take advantage of optimized calculation) usually
+       * evaluates its input expression in evaluateGroupBatch. The optimized calculation doesn't
+       * necessarily work on batches, but input expressions still have to be evaluated, so we take
+       * care of them here.
+       */
+      RowPositionInBatch rp = getPosition(rowNum);
+      evaluator.evaluateInputExpr(bufferedBatches.get(rp.batchIndex));
```

Comment:
Can this cause slowness/overhead on other optimized calculated PTF evaluators which doesn't need this calculation. Is this only need for lead/lag evaluators?

Reply:
I need to rethink this...I can see two problems:

1. sum/avg runs optimized calculation but still ends up calling evaluateGroupBatch which takes care of this part, so not needed (so I would say this is for lead/lag only), so maybe I have to create another method reflecting this property of an evaluator
2. this code evaluates input expression for the same batch which contains the current row, but only for that one, which can be still problematic if a lead function points to a later record, which is present in the next batch for instance...